### ISSUE TITLE
How to avoid BrokenPipeError if a message has been acked (acks_late)

### ISSUE THREAD

**Issue by stanleyli60** [celery/celery#7774]
### Discussed in https://github.com/celery/celery/discussions/7762

<div type='discussions-op-text'>

<sup>Originally posted by **stanleyli60** September 15, 2022</sup>
Hi all,

I am using Celery 5.2.2 + RabbitMQ setup. My task has been configured as acks_late where we only acknowledge the task when finished.

i.e. 
@shared_task(bind=True, acks_late=True)
def execute_sth()

I have multiple celery workers setup with (concurrency = 8 and worker_prefetch_multiplier= 1).

The task is expected to last from 10 min to 20min. What has happened is that the task is delivered to one of the worker and it started working on it. And then in the middle of its execution, the task is re-delivered to another worker and I have code to complete the task right away if it is executed for the second time. So the second worker acks the message earlier. When the first worker is finished and tries to ack on the same message, I am getting a broken pipe error.

Below is some example logs:
2022-09-15 12:14:35  ecs-autoscale-celery-30-autoscale-celery-dad78da2d996eb9b2e00 received 5391871
.
.
.
(first worker received the task and start working on it)
.
.
2022-09-15 12:20:28 ecs-autoscale-celery-30-autoscale-celery-c0e6a4e6e9d1d9d05000 received 5391871
(second worker received the same task and start working on it. And finished right away and ack the message)
.
.
2022-09-15 12:22:15 ecs-autoscale-celery-30-autoscale-celery-dad78da2d996eb9b2e00 succeeded
(first worker finally finished and tries to ack on the message)

Getting the following exception message:
Message
Couldn't ack 206, reason:"BrokenPipeError(32, 'Broken pipe')"

Any idea on how I should avoid this problem?</div>

**Comment by HansAarneLiblik** [comment]
I have the same-ish problem

* Task is sent to queue
* Celery worker picks up the task (
  * ```py 
    @app.task(name="core.debug_delay_test")
    def debug_delay_task() -> None:
        logger.info("Starting")
        time.sleep(600)
        logger.info("Finished")
       ```
* *10 minutes goes by*
* Celery worker crashes

Worker logs:
```
[2022-10-03 13:47:33,612: INFO/MainProcess] Task core.debug_delay_test[c73d6aef-59a4-4578-be15-11ba29b99803] received
[2022-10-03 13:47:33,612: DEBUG/MainProcess] TaskPool: Apply <function fast_trace_task at 0x118f271c0> (args:('core.debug_delay_test', 'c73d6aef-59a4-4578-be15-11ba29b99803', {'lang': 'py', 'task': 'core.debug_delay_test', 'id': 'c73d6aef-59a4-4578-be15-11ba29b99803', 'shadow': None, 'eta': None, 'expires': None, 'group': None, 'group_index': None, 'retries': 0, 'timelimit': [None, None], 'root_id': 'c73d6aef-59a4-4578-be15-11ba29b99803', 'parent_id': None, 'argsrepr': '()', 'kwargsrepr': '{}', 'origin': 'gen16780@TRT-X-L-2778.privatesite.com', 'ignore_result': False, 'compression': 'application/x-gzip', 'properties': {'content_type': 'application/x-python-serialize', 'content_encoding': 'binary', 'application_headers': {'lang': 'py', 'task': 'core.debug_delay_test', 'id': 'c73d6aef-59a4-4578-be15-11ba29b99803', 'shadow': None, 'eta': None, 'expires': None, 'group': None, 'group_index': None, 'retries': 0, 'timelimit': [None, None], 'root_id': 'c73d6aef-59a4-4578-be15-11ba29b99803', 'parent_id': None, 'argsrepr': '()', 'kwargsrepr': '{}', 'origin': 'gen16780@TRT-X-L-2778.privatesite.com', 'ignore_result': False,... kwargs:{})
[2022-10-03 13:47:34,422: DEBUG/MainProcess] (0.003) SELECT "DJANGO_CELERY_RESULTS_TASK5A9B"."ID", "DJANGO_CELERY_RESULTS_TASK5A9B"."TASK_ID", "DJANGO_CELERY_RESULTS_TASK5A9B"."PERIODIC_TASK_NAME", "DJANGO_CELERY_RESULTS_TASK5A9B"."TASK_NAME", "DJANGO_CELERY_RESULTS_TASK5A9B"."TASK_ARGS", "DJANGO_CELERY_RESULTS_TASK5A9B"."TASK_KWARGS", "DJANGO_CELERY_RESULTS_TASK5A9B"."STATUS", "DJANGO_CELERY_RESULTS_TASK5A9B"."WORKER", "DJANGO_CELERY_RESULTS_TASK5A9B"."CONTENT_TYPE", "DJANGO_CELERY_RESULTS_TASK5A9B"."CONTENT_ENCODING", "DJANGO_CELERY_RESULTS_TASK5A9B"."RESULT", "DJANGO_CELERY_RESULTS_TASK5A9B"."DATE_CREATED", "DJANGO_CELERY_RESULTS_TASK5A9B"."DATE_DONE", "DJANGO_CELERY_RESULTS_TASK5A9B"."TRACEBACK", "DJANGO_CELERY_RESULTS_TASK5A9B"."META" FROM "DJANGO_CELERY_RESULTS_TASK5A9B" WHERE "DJANGO_CELERY_RESULTS_TASK5A9B"."TASK_ID" = c73d6aef-59a4-4578-be15-11ba29b99803 FETCH FIRST 21 ROWS ONLY; args=('c73d6aef-59a4-4578-be15-11ba29b99803',); alias=default
[2022-10-03 13:47:34,428: DEBUG/MainProcess] (0.003) INSERT INTO "DJANGO_CELERY_RESULTS_TASK5A9B" ("TASK_ID", "PERIODIC_TASK_NAME", "TASK_NAME", "TASK_ARGS", "TASK_KWARGS", "STATUS", "WORKER", "CONTENT_TYPE", "CONTENT_ENCODING", "RESULT", "DATE_CREATED", "DATE_DONE", "TRACEBACK", "META") VALUES (c73d6aef-59a4-4578-be15-11ba29b99803, None, core.debug_delay_test, "()", "{}", STARTED, worker@TRT-X-L-2778.privatesite.com, application/json, utf-8, {"pid": 17409, "hostname": "worker@TRT-X-L-2778.privatesite.com"}, None0, None1, None, None2) RETURNING "DJANGO_CELERY_RESULTS_TASK5A9B"."ID" INTO None3; args=('c73d6aef-59a4-4578-be15-11ba29b99803', None, 'core.debug_delay_test', '"()"', '"{}"', 'STARTED', 'worker@TRT-X-L-2778.privatesite.com', 'application/json', 'utf-8', '{"pid": 17409, "hostname": "worker@TRT-X-L-2778.privatesite.com"}', Oracle_datetime(2022, 10, 3, 13, 47, 34, 425535), Oracle_datetime(2022, 10, 3, 13, 47, 34, 425581), None, '{"children": []}', <django.db.backends.oracle.utils.InsertVar object at 0x1191ee9e0>); alias=default
[2022-10-03 13:47:34,434: INFO/MainProcess] core.debug_delay_test[c73d6aef-59a4-4578-be15-11ba29b99803]: Executing task: (), {}, priority=3
[2022-10-03 13:47:34,434: INFO/MainProcess] core.debug_delay_test[c73d6aef-59a4-4578-be15-11ba29b99803]: Starting
[2022-10-03 13:57:34,446: INFO/MainProcess] core.debug_delay_test[c73d6aef-59a4-4578-be15-11ba29b99803]: Finished
[2022-10-03 13:57:34,451: DEBUG/MainProcess] (0.003) SELECT "DJANGO_CELERY_RESULTS_TASK5A9B"."ID", "DJANGO_CELERY_RESULTS_TASK5A9B"."TASK_ID", "DJANGO_CELERY_RESULTS_TASK5A9B"."PERIODIC_TASK_NAME", "DJANGO_CELERY_RESULTS_TASK5A9B"."TASK_NAME", "DJANGO_CELERY_RESULTS_TASK5A9B"."TASK_ARGS", "DJANGO_CELERY_RESULTS_TASK5A9B"."TASK_KWARGS", "DJANGO_CELERY_RESULTS_TASK5A9B"."STATUS", "DJANGO_CELERY_RESULTS_TASK5A9B"."WORKER", "DJANGO_CELERY_RESULTS_TASK5A9B"."CONTENT_TYPE", "DJANGO_CELERY_RESULTS_TASK5A9B"."CONTENT_ENCODING", "DJANGO_CELERY_RESULTS_TASK5A9B"."RESULT", "DJANGO_CELERY_RESULTS_TASK5A9B"."DATE_CREATED", "DJANGO_CELERY_RESULTS_TASK5A9B"."DATE_DONE", "DJANGO_CELERY_RESULTS_TASK5A9B"."TRACEBACK", "DJANGO_CELERY_RESULTS_TASK5A9B"."META" FROM "DJANGO_CELERY_RESULTS_TASK5A9B" WHERE "DJANGO_CELERY_RESULTS_TASK5A9B"."TASK_ID" = c73d6aef-59a4-4578-be15-11ba29b99803 FETCH FIRST 21 ROWS ONLY; args=('c73d6aef-59a4-4578-be15-11ba29b99803',); alias=default
[2022-10-03 13:57:34,470: DEBUG/MainProcess] (0.004) UPDATE "DJANGO_CELERY_RESULTS_TASK5A9B" SET "TASK_ID" = c73d6aef-59a4-4578-be15-11ba29b99803, "PERIODIC_TASK_NAME" = NULL, "TASK_NAME" = core.debug_delay_test, "TASK_ARGS" = "()", "TASK_KWARGS" = "{}", "STATUS" = SUCCESS, "WORKER" = worker@TRT-X-L-2778.privatesite.com, "CONTENT_TYPE" = application/json, "CONTENT_ENCODING" = utf-8, "RESULT" = null, "DATE_CREATED" = 2022-10-03 13:47:34.425535, "DATE_DONE" = core.debug_delay_test0, "TRACEBACK" = NULL, "META" = core.debug_delay_test1 WHERE "DJANGO_CELERY_RESULTS_TASK5A9B"."ID" = core.debug_delay_test2; args=('c73d6aef-59a4-4578-be15-11ba29b99803', 'core.debug_delay_test', '"()"', '"{}"', 'SUCCESS', 'worker@TRT-X-L-2778.privatesite.com', 'application/json', 'utf-8', 'null', Oracle_datetime(2022, 10, 3, 13, 47, 34, 425535), Oracle_datetime(2022, 10, 3, 13, 57, 34, 465360), '{"children": []}', 23574); alias=default
[2022-10-03 13:57:34,470: INFO/MainProcess] Task core.debug_delay_test[c73d6aef-59a4-4578-be15-11ba29b99803] succeeded in 600.8486739859218s: None
[2022-10-03 13:57:34,474: CRITICAL/MainProcess] Couldn't ack 1, reason:BrokenPipeError(32, 'Broken pipe')
Traceback (most recent call last):
  File "/Users/hansaarne.liblik/Library/Caches/pypoetry/virtualenvs/privateproject-3TQaMwQn-py3.10/lib/python3.10/site-packages/kombu/message.py", line 128, in ack_log_error
    self.ack(multiple=multiple)
  File "/Users/hansaarne.liblik/Library/Caches/pypoetry/virtualenvs/privateproject-3TQaMwQn-py3.10/lib/python3.10/site-packages/kombu/message.py", line 123, in ack
    self.channel.basic_ack(self.delivery_tag, multiple=multiple)
  File "/Users/hansaarne.liblik/Library/Caches/pypoetry/virtualenvs/privateproject-3TQaMwQn-py3.10/lib/python3.10/site-packages/amqp/channel.py", line 1407, in basic_ack
    return self.send_method(
  File "/Users/hansaarne.liblik/Library/Caches/pypoetry/virtualenvs/privateproject-3TQaMwQn-py3.10/lib/python3.10/site-packages/amqp/abstract_channel.py", line 70, in send_method
    conn.frame_writer(1, self.channel_id, sig, args, content)
  File "/Users/hansaarne.liblik/Library/Caches/pypoetry/virtualenvs/privateproject-3TQaMwQn-py3.10/lib/python3.10/site-packages/amqp/method_framing.py", line 186, in write_frame
    write(buffer_store.view[:offset])
  File "/Users/hansaarne.liblik/Library/Caches/pypoetry/virtualenvs/privateproject-3TQaMwQn-py3.10/lib/python3.10/site-packages/amqp/transport.py", line 347, in write
    self._write(s)
BrokenPipeError: [Errno 32] Broken pipe
[2022-10-03 13:57:34,515: WARNING/MainProcess] consumer: Connection to broker lost. Trying to re-establish the connection...
Traceback (most recent call last):
  File "/Users/hansaarne.liblik/Library/Caches/pypoetry/virtualenvs/privateproject-3TQaMwQn-py3.10/lib/python3.10/site-packages/celery/worker/consumer/consumer.py", line 332, in start
    blueprint.start(self)
  File "/Users/hansaarne.liblik/Library/Caches/pypoetry/virtualenvs/privateproject-3TQaMwQn-py3.10/lib/python3.10/site-packages/celery/bootsteps.py", line 116, in start
    step.start(parent)
  File "/Users/hansaarne.liblik/Library/Caches/pypoetry/virtualenvs/privateproject-3TQaMwQn-py3.10/lib/python3.10/site-packages/celery/worker/consumer/consumer.py", line 628, in start
    c.loop(*c.loop_args())
  File "/Users/hansaarne.liblik/Library/Caches/pypoetry/virtualenvs/privateproject-3TQaMwQn-py3.10/lib/python3.10/site-packages/celery/worker/loops.py", line 97, in asynloop
    next(loop)
  File "/Users/hansaarne.liblik/Library/Caches/pypoetry/virtualenvs/privateproject-3TQaMwQn-py3.10/lib/python3.10/site-packages/kombu/asynchronous/hub.py", line 299, in create_loop
    item()
  File "/Users/hansaarne.liblik/Library/Caches/pypoetry/virtualenvs/privateproject-3TQaMwQn-py3.10/lib/python3.10/site-packages/vine/promises.py", line 160, in __call__
    return self.throw()
  File "/Users/hansaarne.liblik/Library/Caches/pypoetry/virtualenvs/privateproject-3TQaMwQn-py3.10/lib/python3.10/site-packages/vine/promises.py", line 157, in __call__
    retval = fun(*final_args, **final_kwargs)
  File "/Users/hansaarne.liblik/Library/Caches/pypoetry/virtualenvs/privateproject-3TQaMwQn-py3.10/lib/python3.10/site-packages/kombu/message.py", line 128, in ack_log_error
    self.ack(multiple=multiple)
  File "/Users/hansaarne.liblik/Library/Caches/pypoetry/virtualenvs/privateproject-3TQaMwQn-py3.10/lib/python3.10/site-packages/kombu/message.py", line 123, in ack
    self.channel.basic_ack(self.delivery_tag, multiple=multiple)
  File "/Users/hansaarne.liblik/Library/Caches/pypoetry/virtualenvs/privateproject-3TQaMwQn-py3.10/lib/python3.10/site-packages/amqp/channel.py", line 1407, in basic_ack
    return self.send_method(
  File "/Users/hansaarne.liblik/Library/Caches/pypoetry/virtualenvs/privateproject-3TQaMwQn-py3.10/lib/python3.10/site-packages/amqp/abstract_channel.py", line 70, in send_method
    conn.frame_writer(1, self.channel_id, sig, args, content)
  File "/Users/hansaarne.liblik/Library/Caches/pypoetry/virtualenvs/privateproject-3TQaMwQn-py3.10/lib/python3.10/site-packages/amqp/method_framing.py", line 186, in write_frame
    write(buffer_store.view[:offset])
  File "/Users/hansaarne.liblik/Library/Caches/pypoetry/virtualenvs/privateproject-3TQaMwQn-py3.10/lib/python3.10/site-packages/amqp/transport.py", line 347, in write
    self._write(s)
BrokenPipeError: [Errno 32] Broken pipe
[2022-10-03 13:57:34,554: DEBUG/MainProcess] Closed channel #1
[2022-10-03 13:57:34,554: DEBUG/MainProcess] Closed channel #2
```

**Comment by auvipy** [comment]
https://github.com/celery/celery/issues/3773 is this issue related?

**Comment by stanleyli60** [comment]
I've figured out the root cause of the issue. Making a comment here to help people with a potential same problem.

The issue is actually from the following setup: AWS ECS + celery + kombu.

When AWS ECS decides to scale down, or kill a celery task container, it will issue a warm shutdown to the celery docker container.

When a warm shutdown signal is issued, and if we are waiting for a kombu message for that container, AWS ECS will close the socket connection to the RabbitMQ server. It will then lead to a ConnectionResetError exception from the task that is waiting for the Kombu message. And if there is another task running and finishes, it will cause the BrokenPipeError because the connection to RabbitMQ has been closed by AWS ECS.

Hope this helps anyone with the same ish problem.
